contributors/design-proposals/api-machinery/client-go-context.md

```diff
@@ -0,0 +1,153 @@
+and also to allow exposing context based http tracing [go blog on context based httptrace](https://blog.golang.org/http-tracing) to consumers of the client.
+
+## Motivation and Goals
+
+When using client-go, external calls to the Kubernetes API are made in order to manage resources. The initiator of these calls may find due to an external event, 
+that they no longer need the resource they had requested, but they have no means by which to inform the client of this change. 
+
+When using client-go, having the ability to add context based http tracing would be valuable for issues around performance, observability and debugging. 
+
+- Allow consumers of the client to indicate, in an idiomatic way, that the action that caused them to invoke the client has been cancelled.
+- Allow consumers of the client to instrument and gain insight via context based http tracing.
+
+## Non Goals
+
+- Introduce a transactional system for write operations whereby upon cancellation the client would somehow rollback any current write operations.
+
+## API
+
+There are two approaches outlined. The first is a non breaking change that appends to the existing API for resources but perhaps puts more burden on the client consumer. 
+The second is a breaking change to the client-go api, that is potentially more idiomatic and clear.
+ 
+Common to both is the fact that the request type in client-go  already exposes a ```Context```  function: [request context set on client-go rest.Request](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L393). 
+If set, this context is passed all the way through to the underlying ```http.Request```: [request context passed to http request code](https://github.com/kubernetes/client-go/blob/master/rest/request.go#L484). 
+So to achieve the goals, we need only modify the client API to allow passing in an external context that is then used to set the existing context on the request type.
+
+## Chosen Implemenation
+
+### Breaking change
+
+The chosen option is to pass the context with each action on a resource. This may be more idiomatic and explicit: 
+
+```go 
+func (c *pods) Get(ctx context.Context, name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	result = &v1.Pod{}
+	err = c.client.Get().
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Context(c.ctx).
+		Do().
+		Into(result)
+	return
+}
+```
+
+These change would impact on every API. Although the changes required would be many, the simplest approach to maintaining current behavior would be to pass ```context.TODO()```.
+
+For the ```listers``` and ```informers``` the change outlined above (adding new constructors) would still allow a context to be passed through to client functions doing the work.
```

Comment:
Would it make more sense in the context of `listers` and `informers` to specify a context constructor function?

Some sample use cases for context like tracing and deadlines, won't make much sense to set once per `informer` or `lister`, but would make more sense to create a new one every time its needed.

Trace id are commonly set per request and the only thing a deadline set at the construction of the `informer` or `lister` would do is limit the time your program can get resources from the api.